src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -1170,14 +1181,14 @@ public LinkedList(IEnumerable<T> list)
             public void Add(T value)
             {
                 LinkedListNode<T> newNode = new LinkedListNode<T>(value);
-                LinkedListNode<T> last;
+                SpinWait sw = default;
 
-                do
+                while (Interlocked.CompareExchange(ref _last.Next, newNode, null) != null)
                 {
-                    last = _last;
-                } while (!object.ReferenceEquals(Interlocked.CompareExchange(ref _last, newNode, last), last));
+                    sw.SpinOnce();
```

Comment:
@stephentoub probably has more experience designing this type of stuff than I, but I get nervous when I see spin waiting. I am proposing that rather than waiting for another thread to update _last, the current thread is capable of following the Next pointers and determining the final node on its own. Alternately we can use a monitor lock (CompareExchange + spin wait + volatile write is a basic mutex, but lacks the performance optimizations, tuning and diagnostics we invest in monitor)

Reply:
if you feel better, I can just do lock(_first). would you be ok with that?

Reply:
I changed it to use lock instead. please let me know if this is ok. thanks.

Reply:
Yeah, I think that is fine. If we get performance feedback in the future that it is insufficient we can always revisit it then.